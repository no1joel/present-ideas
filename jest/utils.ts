import {
  render as testingLibraryRender,
  RenderOptions,
} from "@testing-library/vue";
import linkify from "vue-linkify";
import { createMemoryHistory, createRouter } from "vue-router";
import routes from "../present_ideas/present_ideas/static/router/routes";
import getStore from "../present_ideas/present_ideas/static/store";

const mockRouter = createRouter({ routes, history: createMemoryHistory() });

export const render = async (component, options = {}) => {
  const store = getStore();
  const defaultRenderOptions: RenderOptions = {
    global: {
      plugins: [store, mockRouter],
      directives: { linkified: linkify },
    },
  };
  const renderOptionsWithDefaults = { ...defaultRenderOptions, ...options };

  const result = testingLibraryRender(component, renderOptionsWithDefaults);

  return result;
};
