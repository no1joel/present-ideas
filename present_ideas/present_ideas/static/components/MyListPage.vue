<template>
  <Layout>
    <h1>This is gonna be your list...</h1>
  </Layout>
</template>

<script>
import Layout from "../layout/Layout.vue";

export default {
  components: { Layout },
};
</script>