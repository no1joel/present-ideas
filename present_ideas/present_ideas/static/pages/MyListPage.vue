<template>
  <div class="container-fluid">
    <div class="row">
      <div class="col">
        <LoadingIndicator v-if="loading" />
        <div class="card-columns">
          <PresentIdeaFormCard
            v-bind:user="$route.params.user"
            v-bind:added-by="$route.params.user"
            v-on:added="fetchMyList"
          />
          <PresentIdeaCard
            v-for="present in presents"
            v-bind="present"
            v-bind:key="
              present.index + present.Thing + present.Price + present.Notes
            "
            v-bind:loading="loading"
            v-on:delete-clicked="deleteClicked"
            v-on:update-name="updateName"
            v-on:update-price="updatePrice"
            v-on:update-notes="updateNotes"
          />
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import LoadingIndicator from "../components/LoadingIndicator.vue";
import PresentIdeaCard from "../components/PresentIdeaCard.vue";
import PresentIdeaFormCard from "../components/PresentIdeaFormCard.vue";
import Layout from "../layout/Layout.vue";

export default {
  components: {
    Layout,
    PresentIdeaCard,
    PresentIdeaFormCard,
    LoadingIndicator,
  },
  data() {
    return { presents: [], loading: false };
  },
  mounted() {
    this.fetchMyList();
  },
  computed: {
    user() {
      return this.$route.params.user;
    },
    baseUrl() {
      return window.location.origin || "http://localhost";
    },
  },
  methods: {
    async fetchMyList() {
      this.loading = true;

      const url = `${this.baseUrl}/api/my_list/${this.user}`;
      const response = await fetch(url);
      const data = await response.json();
      const presents = data.presents;
      this.presents = presents.reverse();

      this.loading = false;
    },
    async deleteClicked({ index }) {
      this.loading = true;
      const url = `${this.baseUrl}/api/delete_idea/`;
      const data = {
        index: index,
        user: this.user,
      };
      await fetch(url, {
        method: "POST",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
        },
        body: JSON.stringify(data),
      });
      this.fetchMyList();
    },
    async updateName({ index, name }) {
      this.loading = true;
      const url = `${this.baseUrl}/api/update_idea_name/`;
      const data = {
        index: index,
        user: this.user,
        name: name,
      };
      await fetch(url, {
        method: "POST",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
        },
        body: JSON.stringify(data),
      });
      this.fetchMyList();
    },
    async updatePrice({ index, price }) {
      this.loading = true;
      const url = `${this.baseUrl}/api/update_idea_price/`;
      const data = {
        index: index,
        user: this.user,
        price,
      };
      await fetch(url, {
        method: "POST",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
        },
        body: JSON.stringify(data),
      });
      this.fetchMyList();
    },
    async updateNotes({ index, notes }) {
      this.loading = true;
      const url = `${this.baseUrl}/api/update_idea_notes/`;
      const data = {
        index: index,
        user: this.user,
        notes,
      };
      await fetch(url, {
        method: "POST",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
        },
        body: JSON.stringify(data),
      });
      this.fetchMyList();
    },
  },
};
</script>
