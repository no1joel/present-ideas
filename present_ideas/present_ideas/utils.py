from typing import Iterable, List

from sheets.spreadsheet import (
    add_row,
    delete_row,
    get_rows,
    get_worksheets,
    set_cell_value,
)


def check_name(name: str) -> bool:
    """Check if the provided name is a sheet."""

    worksheet_names = get_all_names()
    return name.lower() in worksheet_names


def get_all_names() -> Iterable[str]:
    """Return people's names."""

    return (
        worksheet.title.lower()
        for worksheet in get_worksheets()
        if "template" not in worksheet.title.lower()
    )


def get_present_ideas(their_name: str) -> List[dict]:
    # TODO: Embed urls
    _header, *rows = get_rows(their_name)
    presents = [{**row, "index": index} for index, row in enumerate(rows)]
    return presents


def get_row_index(thing_index: int) -> int:
    # TODO: Why is it plus 3?!
    return thing_index + 3
