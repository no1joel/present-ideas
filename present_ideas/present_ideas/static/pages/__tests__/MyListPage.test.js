import MyListPage from "../MyListPage.vue";
import { render } from "./utils";

describe("MyListPage", () => {
  // TODO: Just trying to get it to pass right now.
  // shouldn't mock fetch, shouldn't mess with global
  // Should probably use what vfp used with the api.
  beforeEach(() => {
    const json = async () => ({
      presents: [],
    });
    const response = { json };
    global.fetch = async () => response;
  });
  it("matches snapshot", () => {
    expect(render(MyListPage).container).toMatchSnapshot();
  });
});
