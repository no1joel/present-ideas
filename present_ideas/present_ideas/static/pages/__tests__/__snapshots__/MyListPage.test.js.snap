// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyListPage matches snapshot 1`] = `
<div>
  <div
    class="container-fluid"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <!---->
         
        <div
          class="card-columns"
        >
          <div
            class="card"
          >
            <!---->
             
            <div
              class="card-header"
            >
              <h5
                class="card-title m-0"
              >
                Add New!
              </h5>
               
              <!---->
            </div>
             
            <div
              class="card-body"
            >
              <form
                class="form-horizontal"
              >
                <div
                  class="form-group"
                >
                  <label
                    for="thing-input"
                  >
                    Thing*
                  </label>
                   
                  <input
                    class="form-control"
                    id="thing-input"
                    name="thing"
                    placeholder="Something I want..."
                    type="text"
                  />
                </div>
                 
                <div
                  class="form-group"
                >
                  <label
                    for="price-input"
                  >
                    Price
                  </label>
                   
                  <input
                    class="form-control"
                    id="price-input"
                    name="price"
                    placeholder="Enter a price or a range, e.g. £10-£20"
                    type="text"
                  />
                </div>
                 
                <div
                  class="form-group"
                >
                  <label
                    for="notes-input"
                  >
                    Notes
                  </label>
                   
                  <textarea
                    class="form-control"
                    id="notes-input"
                    name="notes"
                    placeholder="Enter a link to a specific thing, a description of what you want, why you want it, anything!"
                  />
                </div>
                 
                <button
                  class="btn btn-success btn-block"
                  disabled="disabled"
                  type="button"
                >
                  
        Add
      
                </button>
              </form>
            </div>
          </div>
           
        </div>
      </div>
    </div>
  </div>
</div>
`;
